Test vectors

jurraca opened the pull request:
Stacked on https://github.com/asmap/kartograf/pull/32

Two commits:

- first moves all util-style modules to their own folder/namespace
- second adds test vectors/fixtures to test network merging from. Also adds an __init__.py to the tests folder so pytest can resolve modules.

Let me know if this is not ok, I struggled with Python's imports logic a bit.

fjahr commented on kartograf/util/generate_data.py:
```diff
@@ -0,0 +1,105 @@
+
+MAX_ASN = 33521664
+
+def generate_ip(ip_type="v4", subnet_size="16"):
```
How about having these test specific util functions under a util folder in tests, similar to bitcoin core: https://github.com/bitcoin/bitcoin/tree/master/test? If we don't need these in the normal code, I would just put them in the test space and let them be shared there.

jurraca commented:
Cleaned this up a bit.
I reverted the change to `format_pfx` from 97f0fbe552e9c8d4b42c8a03468f47d7bb0fb5c3 , since it just removed the leading zero of private networks and resulted in invalid IP networks ( `0.128.0.0/16` -> `.128.0.0/15`). The follow ups to this will be changing how we check that the prefix we ingest is invalid vs valid but poorly formatted in the parsing logic.

jurraca commented on kartograf/util.py:
```diff
@@ -133,22 +133,32 @@ def wait_for_launch(wait):
         return pfx
 
 
+def is_valid_pfx(pfx):
```
yea, thanks that makes sense. Is there a reason we're checking `ipaddress.ip_address(pfx)` actually? Shouldn't we reject things that are not networks?

fjahr replied:
> Is there a reason we're checking ipaddress.ip_address(pfx) actually?

Hm, not sure I understand the question correctly but the idea in my old code was to use `str` -> `ipaddress` -> `str` to standardize the formatting. Stuff like leading zeros should be handled consistently through this for example. At least that is the rationale I remember from taking a quick look.

fjahr replied:
The idea here is of course that I expected `ipaddress` to handle this better than I ever could by writing a custom parser. I.e. parsing to `ipaddress` should be able to read as many weird formats as possible and then stringifying `ipaddress` should always be consistent. But honestly I can't say that I have put in a lot of effort to verify that `ipaddress` is super solid.

jurraca replied:
I just meant checking `ipaddress.ip_address(pfx)` vs `ipaddress.ip_network(pfx)` -- the first checks an address, the other a network. Don't we only want to check networks?

fjahr replied:
Ah ok, I misunderstood the question. If I remember correctly single IPs are valid entries in RPKI repos and IRR. So there is nothing wrong with them per se although i don't remember if I looked into if there are specific use cases behind it. But this looks like something where I encountered an error and then looked at the spec and then went with what the spec allowed.

jurraca commented on kartograf/util.py:
```diff
@@ -133,22 +133,32 @@ def wait_for_launch(wait):
     """
     try:
         if "/" in pfx:
-            pattern = r"^0+"
```
yep, done

fjahr commented:
> edit: in fact, this [04953b8](https://github.com/asmap/kartograf/pull/37/commits/04953b86c1e34bd7acf5f7ebd6d1993ab72ff37d) commit is a bug fix -- tests are broken. I can split it out on its own if that's better. lmk.

Would have been easier for me to keep it separate. It's not an issue if you open many small PRs if you are explicit which ones you want me to look at first. As I mentioned in the other one you now, you can make the others drafts or say "this depends on X, review that one first".

jurraca commented:
> Would have been easier for me to keep it separate.

yea, sorry about this. Separated the bug fix out to #47 and rebased this branch on that.

fjahr commented:
Looks good to me but needs rebase first

fjahr reviewed (commented):
ACK, leaving some suggestion for a follow-up but merging this for now since it's a good addition either way.
